[CORE-8760] `rptest`: fix race-y checks in `test_index_recovery_after_upgrade`

WillemKauf opened the pull request:
This test can race with segment rolls and un-self-compacted segments, since self compaction may not occur before `redpanda` version changes and node restarts occur. Thus, breaking the expectations around the `mtime()` stats.

## The issues with this test (two points of failure)

To better illustrate why this test is race-y, let's look at parts of the test in chronological order:

This function, `get_closed_segment2mtime()`, is used to gather the `mtime()` statistics for the existing, **closed** segments in storage. 

https://github.com/redpanda-data/redpanda/blob/be699729fbbb0b48cc684d4417383ad1320103b7/tests/rptest/tests/compaction_recovery_test.py#L175-L187

To begin, it is called while the nodes are still running a `redpanda` version with a compaction index footer version `V1`. Call this part of the test version epoch 1.
https://github.com/redpanda-data/redpanda/blob/be699729fbbb0b48cc684d4417383ad1320103b7/tests/rptest/tests/compaction_recovery_test.py#L194-L195

There is already a race condition here, as a segment could be rolled in between  `produce_and_wait_for_compaction()` and `get_closed_segment2mtime()`, thereby appearing in the `mtime` statistics gathered in `seg2mtime_1`, but not self compacted (more on the issues with `produce_and_wait_for_compaction()` later- but for now, take my word that it doesn't guarantee there won't be an open segment that eventually rolls to become a closed segment after it returns in this version epoch). Call this segment `S1` for future reference.

After this, we upgrade and restart the `to_restart` node to a `redpanda` version with a compaction index footer version `V2`. Call this version epoch 2.
https://github.com/redpanda-data/redpanda/blob/be699729fbbb0b48cc684d4417383ad1320103b7/tests/rptest/tests/compaction_recovery_test.py#L198-L200

Now, repeat the process of producing, "waiting" for compaction, and gathering `mtime` statistics. 
https://github.com/redpanda-data/redpanda/blob/be699729fbbb0b48cc684d4417383ad1320103b7/tests/rptest/tests/compaction_recovery_test.py#L202-L206

However, we now assert on some expectations around the gathered `seg2mtime_2` values:

https://github.com/redpanda-data/redpanda/blob/be699729fbbb0b48cc684d4417383ad1320103b7/tests/rptest/tests/compaction_recovery_test.py#L209-L212

### 1. L212 is the first possible point of failure.

`S1` is present in `seg2mtime_1`, and also present in `seg2mtime_2`. It was rolled in version epoch 1, but importantly, **not self compacted**. The result of this is that in version epoch 2, when processing the compaction index footer for `S1`, `redpanda` decides to self compact `S1`, resulting in a rewritten segment `S1'` with a `mtime` that differs from `S1`, **causing the assert to fail**.

But, let's say we avoid this race. There are more issues with the test past this point.

We downgrade the `to_restart` node to the previous version of `redpanda` with a compaction index footer version `V1`, call this version epoch 3.

https://github.com/redpanda-data/redpanda/blob/be699729fbbb0b48cc684d4417383ad1320103b7/tests/rptest/tests/compaction_recovery_test.py#L214-L216

Again, produce and "wait" for compaction.

https://github.com/redpanda-data/redpanda/blob/be699729fbbb0b48cc684d4417383ad1320103b7/tests/rptest/tests/compaction_recovery_test.py#L218-L219

Now, even more asserts on our expected `seg2mtime_3` data:

https://github.com/redpanda-data/redpanda/blob/be699729fbbb0b48cc684d4417383ad1320103b7/tests/rptest/tests/compaction_recovery_test.py#L221-L228

### 2. L228 is the second possible point of failure.

The assert on L225 would have a similar race condition to the one presented in the first point of failure, but L212 will fail first in this case.

L228 is dangerous because there is another race-y issue in this test, caused by the flawed logic in `produce_and_wait_for_compaction()`. More specifically, the flawed logic in `no_big_closed_segments()`, which is used to determine if compaction is "complete".

https://github.com/redpanda-data/redpanda/blob/be699729fbbb0b48cc684d4417383ad1320103b7/tests/rptest/tests/compaction_recovery_test.py#L167-L171

This function tests if any segments are currently closed AND uncompacted. It does not guarantee that all open segments will be closed and compacted by the end of the version epoch.

What this means is that it is possible for another segment, call it `S2`, to have the following sequence of events, starting in version epoch one:
1. (Version Epoch 1) Open at time of `produce_and_wait_for_compaction()`, which returns
2. (Version Epoch 1)  Open at time of `get_closed_segment2mtime()` (therefore doesn't show up in `seg2mtime_1`), which returns
3. (Version Epoch 1) **Closed and self compacted with compaction index footer V1** after both of those function calls, but before the node is restarted
4. (Version Epoch 2) `S2` shows up in `segm2time_2`
5. (Version Epoch 3) The check `if index in seg2mtime_1` on L223 returns `False` for `S2`.
6. (Version Epoch 3) We then assert that `seg2mtime_3[index] > seg2mtime_2[index]`, which fails because in actuality, `S2`'s compaction index was written with footer version `V1`, and therefore won't be rewritten. We intended to hit the assert on L225 for this segment, but did not because of the race condition.


### Whew so what's the fix

There is no good way to guarantee there aren't races around segment rolls, and there also aren't any good ways to guarantee or even check that a segment is self compacted by the end of a version epoch/before a restart or not in this test.

The best way to ensure correctness in this test is to actually parse the version of the compacted index footer and use that to assert on the `mtime` values gathered.

Therefore, to fix the race conditions, **use a newly added compaction index footer reader from `compute_storage.py` to check the version of compacted index footers, and asserting what we can for the `mtime` values based on that version**.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--7ddb6673-01f4-427b-86e4-5b85f4558189-->
<!--TEST RESULTS-->
<!--62d8c570-e846-408f-a74f-47433b375809-->
<!--TEST RESULTS-->
<!--b41cbd73-a5be-4405-903f-530a55041117-->
<!--TEST RESULTS-->
<!--b0c945cd-35dc-41f1-8dc1-b12fa8d0cb05-->
<!--TEST RESULTS-->
<!--87c36af3-2d3c-4900-b09a-2700ecdefb66-->
<!--TEST RESULTS-->
<!--9402bd27-031e-49e3-8eca-aab003f6596d-->
<!--TEST RESULTS-->
<!--6bc50ea2-2400-4166-adf8-f5ac0d1bd5e7-->
<!--TEST RESULTS-->
<!--b3225804-839e-4b07-b90a-c19828bb2895-->
<!--TEST RESULTS-->
<!--78260c01-bb54-4100-99be-b72846af63bb-->
<!--TEST RESULTS-->
<!--f4aec91a-74d9-410b-b290-0b3b5dddd7ee-->
<!--TEST RESULTS-->
<!--b27fa84e-a8d4-4b0d-aaac-73099bd3ebb1-->
<!--TEST RESULTS-->
<!--7440f131-22fa-4003-b7bc-3a14422da026-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62107">build#62107</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62107#019529d5-d4b9-4c50-b297-c88be6119185 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62115">build#62115</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62115#01952ab8-c161-472a-bcec-3c67a7e0e6dd | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62115#01952acd-b310-4402-b4ce-6475c54cd11d | FLAKY | 1/2 |
| rptest.tests.retention_policy_test.ShadowIndexingCloudRetentionTest.test_cloud_time_based_retention.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62115#01952acd-b30f-4995-9089-dc1af7203903 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/62115#01952ab8-c163-4118-b725-5008ec3e1087 | FLAKY | 1/2 |
| storage_e2e_single_thread_rpunit.storage_e2e_single_thread_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/62115#01952a71-6eda-462c-8080-0b4a053566d4 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62124">build#62124</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62124#01952bb0-c90a-468d-b827-13559641f932 | FLAKY | 50/51 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62124#01952bb0-c90a-4bf7-8a90-f52e84cbab5b | FLAKY | 50/53 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62124#01952bb4-2fe3-4789-8f94-86b3428de412 | FLAKY | 50/54 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62124#01952bb4-2fe3-4e3d-9c76-92cd013c08d5 | FLAKY | 50/52 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62138">build#62138</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62138#01952fa4-85a7-40d6-b5fb-aa0026a7ab84 | FAIL | 0/20 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62138#01952fb7-ad7b-4772-89d6-6826dc95188c | FAIL | 0/20 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/62138#01952fa4-85a9-484d-9c2c-85842acf3595 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62144">build#62144</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/62144#01953057-ffec-4306-97b7-31a722671e43 | FLAKY | 1/2 |
| rptest.tests.archival_test.ArchivalTest.test_all_partitions_leadership_transfer.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62144#019530b4-7567-40fb-a2a4-16a3401fa044 | FLAKY | 1/2 |
| rptest.tests.datalake.datalake_e2e_test.DatalakeE2ETests.test_topic_lifecycle.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62144#019530b4-7567-40fb-a2a4-16a3401fa044 | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=False.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62144#0195309f-e020-49c8-ae9c-99f0343ba655 | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=False.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62144#0195309f-e020-4967-bac8-fdaf57e82fe9 | FLAKY | 1/2 |
| storage_e2e_single_thread_rpunit.storage_e2e_single_thread_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/62144#01953057-ffec-4306-97b7-31a722671e43 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62152">build#62152</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62152#01953181-e34d-4210-823a-cf270e1202d5 | FLAKY | 50/55 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62152#01953181-e34d-4c1c-accd-ca1ee5e3e9a5 | FLAKY | 50/55 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62152#0195318f-9dd4-4c96-8c6b-bbaf27608048 | FLAKY | 50/55 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62152#0195318f-9dd5-4336-88e2-4328e2f7d769 | FLAKY | 50/54 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62157">build#62157</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62157#01953337-a5db-43cb-ad95-12f595b93803 | FLAKY | 1/2 |
| rptest.tests.datalake.custom_partitioning_test.DatalakeCustomPartitioningTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62157#01953324-87c9-42ac-b748-168c8b293a6e | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/62157#01953337-a5dc-4615-88a1-ed74e9e2320a | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62167">build#62167</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62167#0195353f-d8b0-406e-877c-9dd7d2b9a102 | FLAKY | 20/24 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62167#0195353f-d8b1-46ce-9dcf-1ceab654743f | FLAKY | 20/22 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62167#0195353f-d8b1-479d-b485-226d96e77409 | FLAKY | 20/23 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62167#0195353f-d8b1-48a9-9d08-5007b25f58f2 | FLAKY | 20/23 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62167#0195353f-d8b1-49c7-bf39-fc9963f61ad7 | FLAKY | 20/24 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62167#0195353f-d8b7-441c-9791-19cf3e28ec10 | FLAKY | 20/23 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62167#0195353f-d8b7-4e3d-ba15-da3bb5c5d191 | FLAKY | 20/23 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62167#0195353f-d8b7-4ecf-bce1-e5837856cecf | FLAKY | 20/28 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62167#0195353f-d8b8-4391-97e4-69aaad76d7c1 | FLAKY | 20/24 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62167#0195353f-d8b8-4e99-b587-f347c3f4cdb3 | FLAKY | 20/22 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62172">build#62172</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_timing_stress_test.CloudStorageTimingStressTest.test_cloud_storage.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/62172#019536bc-3a42-4def-a503-998672666227 | FLAKY | 1/2 |
| rptest.tests.crash_loop_checks_test.CrashLoopChecksTest.test_crash_report_with_signal.signo=Signals.SIGSEGV.signal_shard=1 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62172#019536a2-77aa-4517-9e37-cfac0cdb4805 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62184">build#62184</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62184#01953811-be1c-4654-92e5-eb9e208ed329 | FLAKY | 20/26 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62184#01953811-be1c-47b6-a484-96f6e54fa206 | FLAKY | 20/24 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62184#01953811-be1d-419b-9e63-9824765c4cd8 | FLAKY | 20/23 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62184#01953811-be1d-4a97-8dfe-ad8c06904ec4 | FLAKY | 20/22 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62184#01953811-be1d-4e6e-bfc8-c9d1229a2209 | FLAKY | 20/22 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62184#01953811-be23-47b7-8d5e-e0ccdb558d5f | FLAKY | 20/22 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62184#01953811-be23-4f80-aafd-30edf383a2e8 | FLAKY | 20/23 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62184#01953811-be24-451c-8a88-733ed9ef876f | FLAKY | 20/21 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62184#01953811-be24-49d9-b1cc-a22f6252814a | FLAKY | 20/23 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62184#01953811-be24-4dd0-a593-e649ae40abab | FLAKY | 20/23 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62197">build#62197</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| opfuzz_test_rpunit.opfuzz_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/62197#019538a4-7a4b-4bcc-b075-9b680cf7b215 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62215">build#62215</a></summary>


| test_id                                                                                               | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62215#01953994-f1a3-401e-8472-6e47aa29fba5 | FLAKY         | 20/22    |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62215#01953994-f1a3-4678-a630-93fd84fda886 | FLAKY         | 20/23    |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62215#01953994-f1a3-4d01-8f27-856ed0633bfa | FLAKY         | 20/24    |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62215#01953994-f1a4-42fe-b49e-d7e96d4b943f | FLAKY         | 20/24    |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62215#01953994-f1a4-4b95-89fa-8e12a86deb7c | FLAKY         | 20/22    |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62215#01953994-f1aa-403a-ab48-728c017f33fd | FLAKY         | 20/21    |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62215#01953994-f1aa-406b-9576-078f68147c31 | FLAKY         | 20/23    |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62215#01953994-f1aa-4224-819b-02e0cad07d6d | FLAKY         | 20/24    |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62215#01953994-f1aa-42e1-b90c-6919f75dacb2 | FLAKY         | 20/24    |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62215#01953994-f1aa-4a72-8e4c-0f5f51d72060 | FLAKY         | 20/23    |
</details>
<!--/body-->

WillemKauf commented:
Still flakey. Hmm.

WillemKauf commented:
Force push to:
* Add new condition to `finished_compaction()` to ensure all segments in `partition` data used to evaluate `mtime()` equality/inequality have finished self-compaction in the version epoch in which they were produced.

WillemKauf commented:
/ci-repeat 2
skip-build
skip-units
dt-repeat=50
tests/rptest/tests/compaction_recovery_test.py

WillemKauf commented:
Looks good!

WillemKauf commented:
Still flakey.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/62138-->
<!--02c85b54-57c3-468c-a68e-d468af8d03c9-->
<!--ddcb01ce-7c15-4dba-9c15-aa469e913a4a-->
<!--title-->
## Retry command for [Build#62138](https://buildkite.com/redpanda/redpanda/builds/62138)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/compaction_recovery_test.py::CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade
```
<!--/body-->

WillemKauf commented:
/ci-repeat 2
skip-build
skip-units
dt-repeat=50
tests/rptest/tests/compaction_recovery_test.py

dotnwat reviewed (commented):
> self compaction may not occur before redpanda version changes and node restarts occur.

did you consider changing the test to delay the cluster changes until the desired pre-change state has been reached?

WillemKauf committed:
`rptest`: fix race-y checks in `compaction_recovery_test`

This test can race with segment rolls and un-self-compacted segments,
since self compaction may not occur before `redpanda` version changes
and node restarts occur. Thus, breaking the expectations around the
compacted index `mtime()` stats.

To fix the race conditions, use the newly added compaction index footer
reader from `compute_storage.py` to be able to check the `version` of the
compaction index footer, and assert accordingly on the `mtime` of segments
per the test expectations.

WillemKauf commented:
/ci-repeat 5
skip-redpanda-build
skip-units
dt-repeat=20
tests/rptest/tests/compaction_recovery_test.py

WillemKauf committed:
`rptest`: add `read_compaction_footer` to `compute_storage.py`

Parses the compaction footer from a compacted index for a segment.
This function implements logic similar to that found in
`compacted_index_chunk_reader::load_footer()` to properly read
compacted index footers for either V1 or V2/V3.

WillemKauf commented:
Force push to:
* Add a compaction index footer parser function, `read_compaction_footer()`, to `remote_scripts/compute_storage.py`
* Use the data returned by this function to determine whether a segment has been self compacted or not in `compaction_recovery_test.py`, to properly account for expectations of the test.

WillemKauf commented:
/ci-repeat 5
skip-redpanda-build
skip-units
dt-repeat=20
tests/rptest/tests/compaction_recovery_test.py

WillemKauf commented:
/ci-repeat 5
skip-redpanda-build
skip-units
dt-repeat=20
tests/rptest/tests/compaction_recovery_test.py

WillemKauf commented:
/ci-repeat 5
skip-redpanda-build
skip-units
dt-repeat=20
tests/rptest/tests/compaction_recovery_test.py

WillemKauf commented:
[No flakes in build #62242](https://buildkite.com/redpanda/redpanda/builds/62242). Fixed

dotnwat reviewed (commented):
> since self compaction may not occur before `redpanda` version changes and node restarts occur.

I don't understand the intricacies of the test, but from this description it sounds like a viable solution would be to wait longer before the upgrade happens. to understand why we need more than that I think the description of the issue and the fix needs expanding a bit.

WillemKauf commented:
>I don't understand the intricacies of the test, but from this description it sounds like a viable solution would be to wait longer before the upgrade happens. to understand why we need more than that I think the description of the issue and the fix needs expanding a bit.

Added an explanation of the two race conditions causing failures to the PR cover letter. Hopefully it's clear enough to follow.

dotnwat reviewed (approved):
I'm good with merging this. But I think there is a bigger question about the test: maybe the ducktape test shouldn't be relying on such low-level details like file modification time, and should rather focus on higher level observable outcomes?

WillemKauf commented:
/backport v25.1.x
